src/constructors/lattices.c

```diff
@@ -0,0 +1,299 @@
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&row_start_vector, row_count);
+
+    for (i = 0; i < row_count; i++) {
+        VECTOR(row_lengths_vector)[i] = size - i;
+        VECTOR(row_start_vector)[i] = 0;
+    }
+
+    IGRAPH_CHECK(triangle_lattice(graph, directed, mutual, &row_lengths_vector, &row_start_vector));
+
+    igraph_vector_int_destroy(&row_lengths_vector);
+    igraph_vector_int_destroy(&row_start_vector);
+    IGRAPH_FINALLY_CLEAN(2);
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t triangle_lattice_rectangle_shape(
+    igraph_t *graph, igraph_integer_t size_x, igraph_integer_t size_y,
+    igraph_bool_t directed, igraph_bool_t mutual
+) {
+    igraph_integer_t row_count = size_y;
+    igraph_vector_int_t row_lengths_vector;
+    igraph_vector_int_t row_start_vector;
+    igraph_integer_t i;
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&row_lengths_vector, row_count);
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&row_start_vector, row_count);
+
+    for (i = 0; i < row_count; i++) {
+        VECTOR(row_lengths_vector)[i] = size_x;
+        VECTOR(row_start_vector)[i] = (row_count - i) / 2;
+    }
+
+    IGRAPH_CHECK(triangle_lattice(graph, directed, mutual, &row_lengths_vector, &row_start_vector));
+
+    igraph_vector_int_destroy(&row_lengths_vector);
+    igraph_vector_int_destroy(&row_start_vector);
+    IGRAPH_FINALLY_CLEAN(2);
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t triangle_lattice_hex_shape(
+    igraph_t *graph, igraph_integer_t size_x, igraph_integer_t size_y,
+    igraph_integer_t size_z, igraph_bool_t directed, igraph_bool_t mutual
+) {
+    igraph_integer_t row_count = size_y + size_z - 1;
+    igraph_vector_int_t row_lengths_vector;
+    igraph_vector_int_t row_start_vector;
+    igraph_integer_t i;
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&row_lengths_vector, row_count);
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&row_start_vector, row_count);
+
+    igraph_integer_t row_length = size_x;
+    igraph_integer_t row_start = size_y - 1;
+    igraph_integer_t first_threshold = MIN(size_y - 1, size_z - 1);
+    igraph_integer_t second_threshold = MAX(size_y - 1, size_z - 1);
+    igraph_integer_t sgn_flag = size_y < size_z ? 0 : -1;
+
+    for (i = 0; i < row_count; i++) {
+        VECTOR(row_lengths_vector)[i] = row_length;
+        VECTOR(row_start_vector)[i] = row_start;
+
+        if (i < first_threshold) {
+            row_length++;
+            row_start--;
+        } else if (i < second_threshold) {
+            row_start += sgn_flag;
+        } else {
+            row_length--;
+        }
+    }
+
+    IGRAPH_CHECK(triangle_lattice(graph, directed, mutual, &row_lengths_vector, &row_start_vector));
+
+    igraph_vector_int_destroy(&row_lengths_vector);
+    igraph_vector_int_destroy(&row_start_vector);
+    IGRAPH_FINALLY_CLEAN(2);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \ingroup generators
+ * \function igraph_triangle_lattice
+ * \brief A triangle lattice with the given shape.
+ *
```

Comment:
Please add `\experimental` here. It means that we do not yet commit to the API, so in case we notice that some changes should be made, e.g. while implementing hex lattices, we can still do that.